core/src/main/java/com/datastax/oss/driver/internal/core/metadata/diagnostic/ring/DefaultTokenRingDiagnostic.java

```diff
@@ -0,0 +1,167 @@
+    Preconditions.checkArgument(
+        !consistencyLevel.isDcLocal() || datacenter != null,
+        "datacenter cannot be null for local consistency level " + consistencyLevel);
+    this.keyspace = keyspace;
+    this.consistencyLevel = consistencyLevel;
+    this.datacenter = datacenter;
+    this.tokenRangeDiagnostics = ImmutableSortedSet.copyOf(tokenRangeDiagnostics);
+  }
+
+  @NonNull
+  @Override
+  public KeyspaceMetadata getKeyspace() {
+    return keyspace;
+  }
+
+  @NonNull
+  @Override
+  public ConsistencyLevel getConsistencyLevel() {
+    return consistencyLevel;
+  }
+
+  @Override
+  public Optional<String> getDatacenter() {
+    return Optional.ofNullable(datacenter);
+  }
+
+  @NonNull
+  @Override
+  public SortedSet<TokenRangeDiagnostic> getTokenRangeDiagnostics() {
+    return tokenRangeDiagnostics;
+  }
+
+  @NonNull
+  @Override
+  public Map<String, Object> getDetails() {
+    ImmutableMap.Builder<String, Object> builder = ImmutableMap.builder();
+    addBasicDetails(builder);
+    addTokenRangeDiagnosticDetails(builder);
+    return builder.build();
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (!(o instanceof TokenRingDiagnostic)) {
+      return false;
+    }
+    TokenRingDiagnostic that = (TokenRingDiagnostic) o;
+    return getKeyspace().equals(that.getKeyspace())
+        && getConsistencyLevel().equals(that.getConsistencyLevel())
+        && getTokenRangeDiagnostics().equals(that.getTokenRangeDiagnostics());
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(keyspace, consistencyLevel, tokenRangeDiagnostics);
+  }
+
+  protected void addBasicDetails(ImmutableMap.Builder<String, Object> builder) {
+    builder
+        .put("status", getStatus())
+        .put("keyspace", keyspace.getName().asCql(true))
+        .put("replication", keyspace.getReplication())
+        .put("consistencyLevel", consistencyLevel);
+    if (datacenter != null) {
+      builder.put("datacenter", datacenter);
+    }
+  }
+
+  protected void addTokenRangeDiagnosticDetails(ImmutableMap.Builder<String, Object> builder) {
+    Map<Boolean, List<TokenRangeDiagnostic>> availableAndUnavailable =
+        tokenRangeDiagnostics.stream()
+            .collect(Collectors.partitioningBy(TokenRangeDiagnostic::isAvailable));
+    List<TokenRangeDiagnostic> available =
+        availableAndUnavailable.getOrDefault(true, Collections.emptyList());
+    List<TokenRangeDiagnostic> unavailable =
+        availableAndUnavailable.getOrDefault(false, Collections.emptyList());
+    builder.put("availableRanges", available.size()).put("unavailableRanges", unavailable.size());
+    if (!unavailable.isEmpty()) {
+      addTop10UnavailableRanges(builder, unavailable);
+    }
+  }
+
+  /**
+   * Extracts up to 10 of the most severely unavailable ranges and returns a map of availability
+   * reports keyed by token ranges, formatted for display.
+   */
```

Comment:
Semi-nit: this comment isn't quite right.  This is a void method so it doesn't return anything... it updates the input builder. :)

Reply:
The comment is from a previous version of the method. I'm fine with side effects here, so I'll just change the javadocs.